### PR TITLE
Completions: Add hugging face provider

### PR DESCRIPTION
**Pull request by philipp-spiess**
This PR adds an experimental hugging face provider that we can use to test OSS models.

We currently have an instance with a deployed version for WizardCoder available, so this prompt is based on the StarCoder FIM prompt and seems to work great in my limited tests.

The biggest issue we have with this approach is latency. So we'll need to work on this next.

## Test plan

- Set the following settings:
   ```
  "cody.completions.advanced.provider": "unstable-huggingface",
  "cody.completions.advanced.serverEndpoint": "https://tjjgqj2ub71aikaq.us-east-1.aws.endpoints.huggingface.cloud",
  "cody.completions.advanced.accessToken": "<TOKEN>",
   ```
- Start a completion request:  
Screenshot 2023-06-19 at 12 03 25

### CODE REVIEW

File: client/cody/src/completions/providers/unstable-huggingface.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    constructor(options: ProviderOptions, unstableHuggingFaceOptions: UnstableHuggingFaceOptions) {
+        super(options)
+        this.serverEndpoint = unstableHuggingFaceOptions.serverEndpoint
+        this.accessToken = unstableHuggingFaceOptions.accessToken
+    }
+
+    public async generateCompletions(abortSignal: AbortSignal): Promise<Completion[]> {
+        // TODO: Add context and language
+        const prompt = `<fim_prefix>${this.prefix}<fim_suffix>${this.suffix}<fim_middle>`
```
Comment (valerybugakov) [inline client/cody/src/completions/providers/unstable-huggingface.ts:27]: Would it be helpful to add a relevant docs link here?

File: client/cody/src/completions/providers/unstable-huggingface.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        const request = {
+            inputs: prompt,
+            parameters: {
+                num_return_sequences: 1,
+                max_new_tokens: this.multilineMode === null ? 40 : 128,
```
Comment (valerybugakov) [inline client/cody/src/completions/providers/unstable-huggingface.ts:33]: Why 40 and 128 (comment about it)?

File: client/cody/src/completions/providers/unstable-huggingface.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        })
+
+        try {
+            const data = (await response.json()) as { generated_text: string }[] | { error: string }
+
+            if ('error' in data) {
+                throw new Error(data.error)
+            }
+
+            const completions: string[] = data.map((c: { generated_text: string }) =>
```
Comment (valerybugakov) [inline client/cody/src/completions/providers/unstable-huggingface.ts:60]:
```suggestion
            const completions: string[] = data.map(c =>
```

File: client/cody/src/main.ts [hunk 4]
```diff
@@ -407,26 +408,41 @@ function createCompletionsProvider(
...
-                break
             }
 
             // eslint-disable-next-line @typescript-eslint/no-floating-promises
             webviewErrorMessenger(
                 'Provider `unstable-codegen` can not be used without configuring `cody.completions.advanced.serverEndpoint`. Falling back to `anthropic`.'
             )
+            break
         }
-        default:
```
Comment (valerybugakov) [inline client/cody/src/main.ts:425]: Why did you decide to remove the default case?

File: client/cody/src/main.ts [hunk 5]
```diff
@@ -407,26 +408,41 @@ function createCompletionsProvider(
         codebaseContext
     )
 
-    let providerConfig: ProviderConfig
+    let providerConfig: null | ProviderConfig = null
     switch (config.completionsAdvancedProvider) {
```
Comment (valerybugakov) [inline client/cody/src/main.ts:412]: Extracting the completion provider creation into a standalone function would be helpful.

**Review by valerybugakov — APPROVED** [review]
Awesome work! Looking forward to learning more about the potential latency reduction for this model.

**Review by abeatrix — APPROVED** [review]
Can't wait to try this!

**Pull request merged**